trailmap/pathway_analysis.py

```diff
@@ -0,0 +1,319 @@
+        else:
+            path_flow = nx.maximum_flow_value(H_sg, sources[0], sinks[0])
+            return path_flow
+
+    else:
+        print("Provided pathway has multiple edges. Flow calculation across \
+               multi-edged pathways is currently unsupported. Returning None")
+        return None
+
+
+def find_simple_cycles(G): # pragma: no cover
+    '''finds cycles in a graph and returns them in a list of lists.
+    '''
+    sc = list(nx.simple_cycles(G))
+    return sc
+
+
+def check_if_sublist(path, set_of_steps):
+    '''Checks to see if a pathway contains each element in steps (list)
+    in order. Returns True/False and the position where the steps begin. If
+    False, returns position -1.
+    '''
+    pos = -1
+    sub_list = False
+
+    for i in range(len(path) - len(set_of_steps)+1):
+        if path[i] == set_of_steps[0]:
+            n = 1
+            while (n < len(set_of_steps) and path[i+n] == set_of_steps[n]):
+                n+=1
+            if n == len(set_of_steps):
+                sub_list = True
+                pos = i + len(set_of_steps)
+                break
+
+    return sub_list, pos
+
+
+def roll_cycle(path, cycle):
+    '''Checks if a cycle can be entered from a node in a simple path. If so,
+    finds the last overlapping node and rolls the cycle so it begins with that
+    node. Returns an empty tuple if the cycle does not overlap with the path.
+    '''
+    rolled = ()
+    # find the shared node that appears last in the path
+    for node in path[-1:0:-1]:
+        if node in cycle:
+            rolled = tuple(np.roll(cycle, -cycle.index(node)),)
+            break
+
+    return rolled
+
+
+def insert_cycles(path, rolled_cycles):
+    '''Inserts already-rolled cycles as a tuple into a path
+    '''
+    path=list(path)
+    for rc in rolled_cycles:
+        path.insert(path.index(rc[0]), rc)
+    path_with_cycles = tuple(path)
+    return path_with_cycles
+
+
+def get_pathways_with_cycles(pathways, sc):
+    '''For the given list of pathways and simple cycles, returns a set of
+    pathways that could contain 1 or more cycles
+    '''
+    pathways_with_cycles = set()
+    pathways = list(pathways) # turn into list, which supports indexing
+    # look at all pathways and simple cycles
+    for path in pathways:
+        rolled_cycles = set()
+        for cycle in sc:
+            rolled_cycle = roll_cycle(path, cycle)
+            if rolled_cycle:
+                rolled_cycles.add(rolled_cycle)
+        
+        # record all the pathways that have cycles, insert single cycle
+        if rolled_cycles:
+            pathways_with_cycles.add(insert_cycles(path, rolled_cycles))
+
+    return pathways_with_cycles
+
+
+def find_paths_with_source(pathways, source):
+    '''returns a subset of pathways that contain a given facility as the source
+    '''
+    subset_pathways = set()
+    for path in pathways:
+        if path[0] == source:
+            subset_pathways.add(path)
+
```

Comment:
Not always sure that this is more clear, but list comprehensions can be useful in cases like this
```suggestion
    subset_pathways = set([ path for path in pathways if path[0] == source])
```